### PR TITLE
Adding kpl metric to track the time for oldest future in processing

### PR DESCRIPTION
**Pull request by yatins47**
*Issue #, if available:*

*Description of changes:* Adding metrics for future time.

Old PR Link : https://github.com/awslabs/amazon-kinesis-producer/pull/319

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 1]
```diff
@@ -540,7 +580,9 @@ protected KinesisProducer(File inPipe, File outPipe) {
         
         long id = messageNumber.getAndIncrement();
         SettableFuture<UserRecordResult> f = SettableFuture.create();
-        futures.put(id, f);
+        SettableFutureTracker futuresTracking = new SettableFutureTracker(f, Instant.now());
```
Comment (ashwing) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:583]: nit : futureTracker

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 2]
```diff
@@ -209,8 +246,11 @@ private void onMetricsResponse(Message msg) {
         private <T> SettableFuture<T> getFuture(Message msg) {
             long id = msg.getSourceId();
             @SuppressWarnings("unchecked")
-            SettableFuture<T> f = (SettableFuture<T>) futures.remove(id);
+            SettableFutureTracker futureTracker = futures.remove(id);
+            SettableFuture<T> f = (SettableFuture<T>) futureTracker.getFuture();
+            oldestFutureTrackerHeap.remove(futureTracker);
             if (f == null) {
+                log.error(String.format("Future for message id %s not found", id));
```
Comment (ashwing) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:253]: Can you add a potential reason?

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 3]
```diff
@@ -88,8 +94,32 @@
...
+
+    @AllArgsConstructor
+    private static class SettableFutureTracker {
+        @NonNull
+        @Getter
+        private SettableFuture<?> future;
+        @NonNull
+        @Getter
+        private Instant timestamp;
+    }
```
Comment (ashwing) [inline java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:122]: Replace all these lomboks with `@Value`

**Review by ashwing — COMMENTED** [review]
Leaving comments. Can you also link the previous PR that had some review comments as well?

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/IKinesisProducer.java [hunk 4]
```diff
@@ -18,6 +18,8 @@
 
     int getOutstandingRecordsCount();
 
+    long getOldestRecordTimeInSeconds();
```
[the remark this answers is not in the document]
Comment (yatins47) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/IKinesisProducer.java:21]: Added. Will need to update the java version to be 1.8 as well from 1.7 currently

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 5]
```diff
@@ -135,7 +165,13 @@ public void run() {
                     } else if (m.hasMetricsResponse()) {
                         onMetricsResponse(m);
                     } else {
-                        log.error("Unexpected message type from child process");
+                        // clear the future here as well since the native core has exhausted its retries.
+                        SettableFuture<UserRecordResult> f = getFuture(m);
```
[the remark this answers is not in the document]
Comment (yatins47) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:169]: Done

File: java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java [hunk 6]
```diff
@@ -574,7 +616,27 @@ protected KinesisProducer(File inPipe, File outPipe) {
...
+     * This is time in seconds from the oldest future submitted from {@link #addUserRecord}
+     * that have not finished. This returns -1 if there no pending records in processing state.
+     *
+     * @return The time in seconds since the oldest record is pending to be sent to kinesis endpoint. Returns -1 if
+     * there are no records in processing state.
+     */
+    @Override
+    public long getOldestRecordTimeInSeconds() {
+        if (oldestFutureTrackerHeap.isEmpty()) {
+            return -1;
```
[the remark this answers is not in the document]
Comment (yatins47) [reply java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java:635]: Done

**Review by ashwing — APPROVED** [review]
lgtm. Fixed the title.

**Comment by yatins47** [comment]
Merging in changes after Ashwin's approval

**Pull request merged**